superset-frontend/packages/superset-ui-core/src/color/CategoricalColorScale.ts

```diff
@@ -39,99 +38,172 @@ class CategoricalColorScale extends ExtensibleFunction {
     const sharedColor = sharedColorMap.get(cleanedValue);
-
-    // priority: parentForcedColors > forcedColors > labelColors
-    let color =
-      this.parentForcedColors?.[cleanedValue] ||
-      this.forcedColors?.[cleanedValue] ||
-      sharedColor;
-
-    if (isFeatureEnabled(FeatureFlag.UseAnalagousColors)) {
-      const multiple = Math.floor(
-        this.domain().length / this.originColors.length,
-      );
-      if (multiple > this.multiple) {
-        this.multiple = multiple;
-        const newRange = getAnalogousColors(this.originColors, multiple);
-        this.range(this.originColors.concat(newRange));
+    // priority: forced color (i.e. custom label colors) > shared color > scale color
+    const forcedColor = this.forcedColors?.[cleanedValue] || sharedColor;
+    const isExistingLabel = this.sliceMap.has(cleanedValue);
+    let color = forcedColor || this.scale(cleanedValue);
+
+    // a forced color will always be used independently of the usage count
+    if (!forcedColor && !isExistingLabel) {
+      if (isFeatureEnabled(FeatureFlag.UseAnalagousColors)) {
+        this.incrementColorRange();
       }
-    }
-    const newColor = this.scale(cleanedValue);
-    if (!color) {
-      color = newColor;
-      if (isFeatureEnabled(FeatureFlag.AvoidColorsCollision)) {
-        this.removeSharedLabelColorFromRange(sharedColorMap, cleanedValue);
-        color = this.scale(cleanedValue);
+      if (
+        // feature flag to be deprecated (will become standard behaviour)
+        isFeatureEnabled(FeatureFlag.AvoidColorsCollision) &&
+        this.isColorUsed(color)
+      ) {
+        // fallback to least used color
+        color = this.getNextAvailableColor(color);
       }
     }
 
-    sharedLabelColor.addSlice(cleanedValue, color, sliceId);
+    // keep track of values in this slice
+    this.sliceMap.set(cleanedValue, color);
+
+    // store the value+color in the shared context
+    if (sliceId) {
+      this.sharedColorMapInstance.addSlice(cleanedValue, color, sliceId);
+    }
 
     return color;
   }
 
   /**
-   * Enforce specific color for given value
+   *
+   * @param color
+   * @returns whether the color is used in this slice
+   */
+  isColorUsed(color: string): boolean {
+    return this.getColorUsageCount(color) > 0;
+  }
+
+  /**
+   * Get the count of the color usage in this slice
+   * @param sliceId
+   * @param color
+   * @returns the count of the color usage in this slice
+   */
+  getColorUsageCount(currentColor: string): number {
+    let count = 0;
+    this.sliceMap.forEach(color => {
+      if (color === currentColor) {
+        count += 1;
+      }
+    });
+    return count;
+  }
+
+  /**
+   * Lower chances of color collision by returning the least used color
+   * Checks across colors of current slice within SharedLabelColorSingleton
+   *
+   * @param currentColor
+   * @returns the least used color that is not the excluded color
+   */
+  getNextAvailableColor(currentColor: string) {
```

Comment:
This is meant to help with color collisions. It orders the colors based on their usage count in the chart. Least used colors will be preferred. This does not guarantee full protection against color collisions, which should be implemented at the individual viz based on their shape.